pkg/extkingpin/path_content_reloader.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package extkingpin
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+)
+
+type fileContent interface {
+	Content() ([]byte, error)
+	Path() string
+}
+
+// PathContentReloader starts a file watcher that monitors the file indicated by fileContent.Path() and runs
+// reloadFunc whenever a change is detected.
+// A debounce timer can be configured via opts to handle situations where many "write" events are received together or
+// a "create" event is followed up by a "write" event, for example. Files will be effectively reloaded at the latest
+// after 2 times the debounce timer. By default the debouncer timer is 1 second.
+// To ensure renames and deletes are properly handled, the file watcher is put at the file's parent folder. See
+// https://github.com/fsnotify/fsnotify/issues/214 for more details.
+func PathContentReloader(ctx context.Context, fileContent fileContent, logger log.Logger, reloadFunc func(), opts ...reloaderOption) error {
```

Comment:
What is the difference between this implementation and the one we have in https://github.com/thanos-io/thanos/blob/main/pkg/receive/config.go? Would be cool if we could have one implementation and reuse it.

Reply:
The implementation in the `receive` package:

- Doesn't have a debouncer to avoid extra reloads when many events are received in a short period of time (i.e. a big enough file write in some systems will trigger many `write` events according to fsnotify's docs).
- Does extra work when a rename happens (a `rename` event is generated for the "old file", then a `write` event is generated for the "new file").
- Has bad logic on: `event.Op^(fsnotify.Chmod|fsnotify.Remove) == 0`. i.e. Imagine `const Chmod = 0b01` and `const Remove = 0b10`. If `var Op = 0b01` (a chmod), the conditional evaluates to: `0b01 ^ (0b01 | 0b10) = 0b01 ^ 0b11 = 0b10 = 2`, which means the chmod actually ends up triggering a configuration reload. I could detect this issue and ensure it won't be broken with tests.

As I mentioned in the PR description, if this works well enough it can be extracted into https://github.com/efficientgo/tools/tree/main/extkingpin and imported here. And we can follow up reusing the code in the `receive` package too. Given fsnotify is full of gotchas and not simple at all, I prefer to gradually introduce the new function, gather & handle feedback, test it across many envs (including production) in a non-critical configuration, and then finally apply it everywhere.

Reply:
I'm all for improving on existing solution, it still makes me wonder whether it is worth to introduce third different implementation (not to mention these all seem to closely follow upstream Prometheus file discovery watcher - https://github.com/prometheus/prometheus/blob/main/discovery/file/file.go#L235-L236 - so if there are serious flaw in the design, we should propose changes there as well).

Ultimately, as you say, `fsnotify` is not the simplest to work for, so I think we should err on the side of having reloads triggered on more events rather than trying to filter them out. Ideally, we should really be reloading config only if we [detect change](https://github.com/thanos-io/thanos/blob/main/pkg/receive/config.go#L231) in content and even then, it's not like these config files are reloaded with very high frequency (so we can afford to "duplicate" a reload, in the worst case).

TL;DR: If we think the added benefits are worth, I'm don't mind adding this implementation, but then it would be nice to track this with an issue (so we eventually bring all three implementations together) and possibly also bring this upstream. Alternatively, solving this as a separate stream of work (i.e. unifying all implementations across Thanos and replacing  them with `extkingpin` implementation) would also be nice, separately from this PR.

Reply:
> Ultimately, as you say, fsnotify is not the simplest to work for, so I think we should err on the side of having reloads triggered on more events rather than trying to filter them out.

I prefer to be sure the implementation is correct when it comes to usage in Thanos Receive or any other high throughput loop/project. Reloading the configuration is relatively slow work and requires some synchronization. This can affect performance considerably if there are many unnecessary reloads.

> Ideally, we should really be reloading config only if we [detect change](https://github.com/thanos-io/thanos/blob/main/pkg/receive/config.go#L231) in content and even then, it's not like these config files are reloaded with very high frequency (so we can afford to "duplicate" a reload, in the worst case).

A duplicate reload is not the worst case, there can be way more than 2 events that end up triggering a reload if the code is not correctly filtering things out. 

The hash checking to only reload config on content changes can be implemented as part of a `reloadFunc`, at discretion `PathContentReloader`'s caller. This way double reads can be avoid and callers can pick their preferred hashing function.

> Alternatively, solving this as a separate stream of work (i.e. unifying all implementations across Thanos and replacing them with extkingpin implementation) would also be nice, separately from this PR.

This is exactly the plan. I plan to get there step by step, with small PRs and time to test things out in real usage.

Reply:
@matej-g I updated the description of the PR to mention the consolidation of implementations, migration to efficientgo/extkingpin, and upstreaming of fixes/improvements into Prometheus. When this is merged I'll open issues for all this work.

Reply:
You dropped this :arrow_right: : :crown: